### ISSUE TITLE
Some @types dependencies are not installed

### ISSUE THREAD

**Issue by perqin** [pnpm/pnpm#4920]
### pnpm version:

7.1.9

### Code to reproduce the issue:

Use this package.json and run `pnpm install`:

```json
{
  "name": "work-log-calendar",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "react": "^18.1.0",
    "react-dom": "^18.1.0",
    "typescript": "^4.7.3",
    "web-vitals": "^2.1.4"
  },
  "packageManager": "pnpm@7.1.9",
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@testing-library/jest-dom": "^5.16.4",
    "@testing-library/react": "^13.3.0",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^28.1.1",
    "@types/node": "^17.0.41",
    "@types/react": "^18.0.12",
    "@types/react-dom": "^18.0.5",
    "@typescript-eslint/parser": "^5.29.0",
    "eslint": "^8.18.0",
    "eslint-config-react-app": "^7.0.1",
    "react-scripts": "5.0.1"
  }
}

```

### Expected behavior:

`@testing-library/jest-dom` has a dependency of `"@types/testing-library__jest-dom": "^5.9.1"`, so `@types/testing-library__jest-dom` should also be installed.

### Actual behavior:

`@types/testing-library__jest-dom` isn't installed:

![image](https://user-images.githubusercontent.com/7088631/175506870-005a2e81-f1f3-4c6b-ac5b-95d12b28870a.png)

### Additional information:

 - `node -v` prints: v14.19.2
 - Windows, macOS, or Linux?: Windows 11
 - If creating an empty package.json and run `pnpm install -D @testing-library/jest-dom`, `@types/testing-library__jest-dom` will be installed.

**Comment by mcmxcdev** [comment]
I just tried to reproduce this: the types library you are referring to ends up under `node_modules/.pnpm/@types+testing-library__jest-dom@5.14.5`

Can you confirm this?

**Comment by dlehmhus** [comment]
I noticed something similar. In my case, the dependency chain starts with `@storybook/testing-library` -> `@testing-library/jest-dom` -> `@types/testing-library__jest-dom`, but the issue is basically the same. The types only get installed to `node_modules/.pnpm/@types+testing-library__jest-dom@5.14.5` and typescript is not happy with that. When I do the install with pnpm v6 the types are also linked to `node_modules/@types/testing-library__jest-dom` and everything works as expected.

**Comment by mcmxcdev** [comment]
Actually, you might be experiencing something that I fixed today. If you're using e.g toBeInDocument() and it doesn't get recognized, put testing-library/jest-dom into types array in your tsconfig, then it works as expected.

**Comment by dlehmhus** [comment]
`.toBeInDocument` was exactly my use case. Putting `testing-library/jest-dom` into the types array doesn't work in my case since I only have `@storybook/jest` in my dependencies. Which internally depends on `testing-library/jest-dom`. Installing and adding `@types/testing-library__jest-dom` directly to the tsconfig works though.

**Comment by mcorbelli** [comment]
any update?

**Comment by ConcernedHobbit** [comment]
This seems like a regression that is affecting everyone with `testing-library/jest-dom` and using its functions.

**Comment by zkochan** [comment]
`@types` are not hoisted by default in pnpm v7: https://github.com/pnpm/pnpm/pull/4459

If you need to hoist types, use this setting in `.npmrc`:

```
public-hoist-pattern[]=*types*
```

**Comment by perqin** [comment]
> `@types` are not hoisted by default in pnpm v7: #4459
> 
> If you need to hoist types, use this setting in `.npmrc`:
> 
> ```
> public-hoist-pattern[]=*types*
> ```

Thanks, this solve my problem! I use `@types*` instead of `*types*` to avoid wrongly hoisting other non-typing packages.

Also, setting this in .npmrc seems to override the default config (`['*eslint*', '*prettier*']`), so my final .npmrc results in:

```ini
public-hoist-pattern[]=*eslint*
public-hoist-pattern[]=*prettier*
public-hoist-pattern[]=@types*
```

**Comment by itayperry** [comment]
It's important to mention what `pnpm` docs say about `public-hoist-pattern`:

"This setting is useful when dealing with some flawed pluggable tools that don't resolve dependencies properly."

https://pnpm.io/npmrc#public-hoist-pattern

**Comment by zkochan** [comment]
I don't know if this issue is related. "flawed pluggable tool that don't resolve dependencies properly" means that the tools don't use node's API to resolve modules. They try to implement their own resolution algorithm that doesn't adhere to the rules used by Node.js. For instance, they might preserve symlinks.

**Comment by codepunkt** [comment]
I'm trying to understand why TypeScript is not picking up a `@types/*` `devDependency` of one of our dev dependencies (`@testing-library/jest-dom` in this case). Is adding "@types*" or "*types*" or anything like that to `public-hoist-patterns` the preferred way of using pnpm in TypeScript code bases?

**Comment by benzittlau** [comment]
@codepunkt I was also wondering this, so I've done some digging to try and understand what's happening here.  I'm not an expert on `pnpm` or `Typescript`, so take this with the appropriate pile of salt given that this is based on only a few hours investigation, and should be confirmed by those who know these technologies better.  Disclaimers out of the way, here's what I understand:

Typescript loads all `@types/*` it finds relative to the directory is is being executed, by crawling up the tree and finding any `node_modules/@types/*` folders, and loading them.  From the Typescript [docs](https://www.typescriptlang.org/tsconfig#types):

> By default all visible ”@types” packages are included in your compilation. Packages in node_modules/@types of any enclosing folder are considered visible. For example, that means packages within ./node_modules/@types/, ../node_modules/@types/, ../../node_modules/@types/, and so on.

The way that `pnpm` works is by installing all packages for a given monorepo into a folder located in the root of the monorepo at `node_modules/.pnpm`.  Within that folder it uses it's own naming convention to place each version of a package within a respective subfolder, which it wraps in a `node_modules` folder in which it can install it's dependencies so they can be found when imported from within the package.  These per-package subfolders are then hardlinked each workspaces folder.  For a concrete example, I created a test project called `apps/testHosting`, with `@testing-library/jest-dom` listed as a dependency in it's`apps/testHosting/package.json` file.  This had the following result:

In the root folder of the monorepo, a folder was created in `.pnpm` for the matching version of that dependency, as well as for that packages typings package, which it lists as a dependency:

`node_modules/.pnpm/@testing-library+jest-dom@5.16.5`
`node_modules/.pnpm/@types+testing-library__jest-dom@5.14.5`

Within the `node_modules` folder for `@testing-library+jest-dom@5.16.5` a hardlink is created to it's @types package (as it is listed as a dependency), which in a situation of typical import node_modules hiearchy crawling would allow it to be found:

```
node_modules/.pnpm/@testing-library+jest-dom@5.16.5/node_modules/@types -> ../../../@types+testing-library__jest-dom@5.14.5/node_modules/@types/testing-library__jest-dom
```

Unfortunately when Typescript crawls for types to load, it never navigates through that path, as it crawls upwards in the hierarchy from the workspace's folder, hitting:

`apps/testHoisting/node_modules/@types`
`apps/node_modules/@types`
`node_modules/@types`

But never hitting:

`node_modules/.pnpm/@testing-library+jest-dom@5.16.5/node_modules/@types`

As far as I can tell, this is an issue that will be encountered with *any* package that lists it's @types as a dependency on another package.  In contract some other libraries (for example, [momentjs](https://www.npmjs.com/package/moment?activeTab=code)) include their typings directly in the package itself, so will work fine in a `pnpm` monorepo.

Given all of the above, for any package that is managing the loading of its types the way that `@testing-library/jest-dom` is, some form of workaround needs to be used to place the respective `@types` package at one of the paths that Typescript will crawl.  This is what's being accomplished with the solution of utilizing the `public-hoist-pattern` configuration, which results in there being a hardlink to the types package placed in the root `node_modules/@types` folder, where Typescript can find it.

There are some downsides to hoisting *all* `@types` packages within a monorepo, though that will be the most blanket solution to handle this for any other library that is using a similar pattern.  Personally I think I will be explicitly including the necessary types as part of the workspace's `package.json` for situations like this.  This is what I was able to put into the `package.json` of my test project to resolve this issue without having to resort to public hoisting:

``` json
  "dependencies": {
    "@testing-library/jest-dom": "^5.16.5",
    "@types/testing-library__jest-dom": "*",
  },
```

**Comment by junepark202012** [comment]
@benzittlau Had this problem with 'topojson-client' and '@types/topojson-specification'
Thanks for the detailed analysis.

**Comment by AlainBartmanDilaw** [comment]
Installing specific @types/testing-library__jest-dom with pnpm works for me !

**Comment by mrbinky3000** [comment]
Got here from a Google search of the issue.  For me, I commented out every line in my tsconfig.json file one at a time until the error went away.  In an foolish effort to exclude tests from the final bundle, I added this to my list of "excludes"

```
  "exclude": [
    // ... other excludes listed here
    "**/*.test.*", // 👈 this was the culprit.
   ]
```

A couple of things wrong with this.  It's not needed. TypeScript won't bundle tests into my production code  unless I import a test into a non-test for some reason.  Also, my tests are written in TypeScript so, not sure what I was trying to do here.  I think since all the tests were ignored, the types were not imported or tree-shaken out or something along those lines.
